persist/fs/commitlog/reader.go

```diff
@@ -109,58 +198,265 @@ func (r *reader) Read() (
+		r.metadata.RLock()
+		metadata, hasMetadata = r.metadata.lookup[entry.Index]
+		r.metadata.RUnlock()
+
+		// The required metadata hasn't been processed yet
+		if !hasMetadata {
+			r.metadata.Lock()
+			metadata, hasMetadata = r.metadata.lookup[entry.Index]
+			if !hasMetadata {
+				pendingMetadata = &readerPendingSeriesMetadataResponse{}
+				pendingMetadata.wg.Add(1)
+				r.metadata.pending[entry.Index] = append(r.metadata.pending[entry.Index], pendingMetadata)
+			}
+			r.metadata.Unlock()
 		}
-		ok = true
 
-		r.metadataLookup[entry.Index] = metadata
+		readResponse.pendingMetadata = pendingMetadata
+
+		// Metadata may or may not actually be nonzero here, the consumer in Read()
+		// will wait if so
+		readResponse.series = metadata
+		readResponse.datapoint = ts.Datapoint{
+			Timestamp: time.Unix(0, entry.Timestamp),
+			Value:     entry.Value,
+		}
+		readResponse.unit = xtime.Unit(byte(entry.Unit))
+		// Copy annotation to prevent reference to pooled byte slice
+		if len(entry.Annotation) > 0 {
+			readResponse.annotation = append([]byte(nil), entry.Annotation...)
+		}
+		// DecRef delayed until this point because even after decoding, underlying
+		// byte slice is still referenced by metadata and annotation
+		arg.bytes.DecRef()
+		arg.bytes.Finalize()
+		r.writeToOutBuf(outBuf, readResponse)
 	}
+	r.metadata.Lock()
+	r.metadata.numBlockedOrFinishedDecoders++
+	// If all of the decoders are either finished or blocked then we need to free
+	// any pending waiters. This also guarantees that the last decoderLoop to
+	// finish will free up any pending waiters (and by then any still-pending
+	// metadata is definitely missing from the commitlog)
+	if r.metadata.numBlockedOrFinishedDecoders >= r.numConc {
+		r.freeAllPendingWaiters()
+	}
+	r.metadata.Unlock()
+	// Close the outBuf now that there is no more data to produce to it
+	close(outBuf)
+}
 
-	if !ok {
-		resultErr = errCommitLogReaderMissingLogMetadata
-		return
+func (r *reader) writeToOutBuf(outBuf chan<- readResponse, readResponse readResponse) {
+	// Guard against all of the decoder loops having full outBuffers (deadlock).
+	// This could happen in the scenario where the commitlog is missing metadata
+	// and the .Wait() call in Read() is blocked waiting for data that will
+	// never arrive and thus is unable to empty the outBufs.
+	// In that case, its clear that the commitlog is malformatted (missing
+	// metadata) so we signal that the Read() call should proceed.
+	for {
+		select {
+		// Happy path, outBuf is not full and our write succeeds immediately
+		case outBuf <- readResponse:
+			return
+		// outBuf is full
+		default:
+			r.metadata.Lock()
+			// Check if all the other decoderLoops are finished or blocked too
+			if r.metadata.numBlockedOrFinishedDecoders >= r.numConc-1 {
+				// If they are, then performing a blocking write would cause deadlock
+				// so we free all pending waiters
+				r.freeAllPendingWaiters()
+				r.metadata.Unlock()
```

Comment:
feels sketch to be allowing this to fall thru and loop intentionally.

Reply:
Are you worried about too much unnecessary looping (could be solved with a short sleep?) or you just don't like the code structure?

This is the only way I know how to write a potentially blocked channel without deadlocking

Reply:
lets chat in person, have a couple of concerns about it

Reply:
cool lets do it